File: Leetcode/0881 Boats to Save People.ts
// O(n log n) time sorting + two pointer solution
function numRescueBoats(people: number[], limit: number): number {
    // O(n log n)
    people.sort((a, b) => a - b);
    let i = 0, j = people.length - 1, nBoats = 0;
    // O(n)
    while(i <= j) {
        const totalWeight = people[i] + people[j];
        if(totalWeight <= limit) {
            i++;
        }
        nBoats++;
        j--;
    }
    return nBoats;
}

File: Leetcode/0881 Boats to Save People.py
class Solution:
    # O(n log n) time sorting + two pointer solution
    def numRescueBoats(self, people, limit):
        people.sort()
        i, j, n_boats = 0, len(people) - 1, 0
        while i <= j:
            total_weight = people[i] + people[j]
            if total_weight <= limit:
                i += 1
            n_boats += 1
            j -= 1
        return n_boats
